datafusion/core/src/execution/context.rs

```diff
@@ -534,6 +534,13 @@ impl SessionContext {
                 let physical = DataFrame::new(self.state(), input);
 
                 let batches: Vec<_> = physical.collect_partitioned().await?;
+                let schema = {
```

Comment:
This solves the bug but feels like the input.schema() should already be correct. We shouldn't need to reach into the record batch to find out the true schema

Reply:
I haven't investigated enough to know what is the "correct" fix.

Reply:
I agree `input.schema()` should be fixed (in the sense that in the `RecordBatch`es produced don't match the schema declared) that is the root cause of the issue 

Perhaps this is related to https://github.com/apache/arrow-datafusion/issues/7636 that you filed

Reply:
See also https://github.com/apache/arrow-datafusion/issues/7636#issuecomment-1737225422

Reply:
This was my initial and quick attempt at fixing #7636 before I dove deeper toward a more correct solution. I'm going to close this out since it should be replaced by https://github.com/apache/arrow-datafusion/pull/7638